lalrpop-test/src/main.rs

```diff
@@ -281,6 +287,65 @@ fn error_test1() {
 }
 
 #[test]
+fn error_recovery_eof() {
+    let errors = RefCell::new(vec![]);
+    util::test(|v| error_recovery::parse_Item(&errors, v), "--", '!'.to_string());
+
+    assert_eq!(errors.borrow().len(), 1);
+    assert_eq!(errors.borrow()[0], ParseError::UnrecognizedToken {
+        token: None,
+        expected: vec![],
+    });
+}
+
+#[test]
+fn error_recovery_eof_without_recovery() {
+    let errors = RefCell::new(vec![]);
+    let tokens = util::tok::tokenize("-").into_iter().map(|t| t.1);
+    let result = error_recovery::parse_Item(&errors, tokens);
+    assert_eq!(result, Err(ParseError::UnrecognizedToken {
+        token: None,
+        expected: vec![],
+    }));
+}
+
+#[test]
+fn error_recovery_extra_token() {
+    let errors = RefCell::new(vec![]);
+    util::test(|v| error_recovery::parse_Item(&errors, v), "(++)", "()".to_string());
+
+    assert_eq!(errors.borrow().len(), 1);
+    assert_eq!(errors.borrow()[0], ParseError::UnrecognizedToken {
+        token: Some(((), Tok::Plus,())),
+        expected: vec![],
+    });
+}
+
+#[test]
+fn error_recovery_dont_drop_unrecognized_token() {
+    let errors = RefCell::new(vec![]);
+    util::test(|v| error_recovery::parse_Item(&errors, v), "(--)", "!".to_string());
```

Comment:
That's just due to how lazily I implemented the parser https://github.com/nikomatsakis/lalrpop/pull/160/files#diff-a8585f30f0fedc26b330b3798c5c7febR22. I could change it to to wrap the result in parentheses.